### PR TITLE
ZoF support

### PR DESCRIPTION
**Pull request by sonicaj**
This PR aims to bring changes to py-libzfs so that we support ZoL(oF) as well. The changes would be split on two levels:

1) Autoconf
2) Cython

Ticket: #78123

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -11,46 +12,64 @@ define(SECTION_TITLE,
...
 
 
-AC_DEFINE(NEED_SOLARIS_BOOLEAN)
-AC_DEFINE(_XPG6)
+dnl We allow setting a zof variable which the users can set to force autoconf to consider the system is zof based
+
+AC_ARG_VAR(zof, AS_HELP_STRING([Force configure to run ZoL(f) checks]))
+AC_ARG_VAR(freebsd_src, AS_HELP_STRING([Provide path for freebsd source tree. This defaults to /usr/src/]))
+
+if [[ -f "/usr/local/include/libzfs/libzfs.h" ]] || [[ "$zof" = true ]]; then
```
Comment (william-gr) [inline configure.ac]: Lets not hardcode /usr/local anywhere here, please.

Lets use an optional --prefix in configure script, defaulting to /usr/local

There are no guarantees everyone uses /usr/local, plus having it configurable makes it easier to support ZoL.

File: libzfs.pyx [hunk 2]
```diff
@@ -2411,23 +2433,41 @@ cdef class ZFSObject(object):
...
+
+            if ret != 0:
+                raise self.root.get_error()
 
-        self.root.write_history('zfs rename', '-f' if forceunmount else '', '-u' if nounmount else '', self.name)
+            self.root.write_history('zfs rename', '-f' if forceunmount else '', '-u' if nounmount else '', self.name)
+
+    ELSE:
+
+        def rename(self, new_name, force_unmount=False, recursive=False):
```
Comment (william-gr) [inline libzfs.pyx]: force_unmount is different from the other function define.

How about we use the same method scope but raise a RuntimeError if `nounmount` is used without HAVE_RENAMEFLAGS_T ?

File: libzfs.pyx [hunk 3]
```diff
@@ -2705,22 +2745,23 @@ cdef class ZFSDataset(ZFSObject):
         if err != 0:
             raise self.root.get_error()
 
-    def get_send_progress(self, fd):
-        cdef zfs.zfs_cmd_t cmd
-        cdef int ret
+    IF HAVE_ZFS_IOCTL_HEADER:
```
Comment (william-gr) [inline libzfs.pyx]: I think I would rather have this method always defined but raise a NotImplementedError if not available.

**Review by william-gr — CHANGES_REQUESTED** [review]
Nice job, I have just a few concerns.

File: libzfs.pyx (hunk 2 above)
Comment (sonicaj) [reply libzfs.pyx]: @william-gr i did that initially but then reverted to this. My reasoning at the time was that the function signature would show that we still accept nounmount ( even if we raise an exception ) - so i reverted to this case. Do you think that is okay and i can move back to having the same signature and raising an exception ?

Comment (william-gr) [reply libzfs.pyx]: Yes, I think that makes things simpler in the long run.

Comment (sonicaj) [reply libzfs.pyx]: Perfect, updating asap

File: configure.ac [hunk 4]
```diff
@@ -11,46 +12,67 @@ define(SECTION_TITLE,
...
 
-AC_DEFINE(NEED_SOLARIS_BOOLEAN)
-AC_DEFINE(_XPG6)
+dnl We allow setting a zof variable which the users can set to force autoconf to consider the system is zof based
+
+AC_ARG_VAR(zof, AS_HELP_STRING([Force configure to run ZoL(f) checks]))
+AC_ARG_VAR(freebsd_src, AS_HELP_STRING([Provide path for freebsd source tree. This defaults to /usr/src/]))
+AC_ARG_VAR(zof_src, AS_HELP_STRING([Provide path for freebsd source tree. This defaults to /usr/local/]))
+
+zof_src=${zof_src:=/usr/local}
```
Comment (william-gr) [inline configure.ac]: But thats not zof_src, thats PREFIX,  to where headers are installed. See previous comment about --prefix in configure, thats common practice.

**Review by william-gr — CHANGES_REQUESTED** [review]

**Review by william-gr — APPROVED** [review]

**Pull request merged**